### PR TITLE
Check ALPN when receiving early data

### LINKED ISSUE #6355: TLS 1.3 server: Add ALPN check when accepting early data
This issue tracks the addition of ALPN check when accepting early data for a PSK established via a NewSessionTicket message.

SPEC: 
"In order to accept early data, the server MUST have accepted a PSK cipher suite and selected the first key offered in the client’s "pre_shared_key" extension. In addition, it MUST verify that the following values are the same as those associated with the selected PSK:
- The TLS version number
- The selected cipher suite
- The selected ALPN [RFC7301] protocol, if any

These requirements are a superset of those needed to perform a 1-RTT handshake using the PSK in question. For externally established PSKs, the associated values are those provisioned along with the key. For PSKs established via a NewSessionTicket message, the associated values are those negotiated in the connection during which the ticket was established.
"

- [ ] Add an `alpn_chosen` or `alpn` field to the session structure. Currently there is such a field in the `mbedtls_ssl_context` structure but we need it at the session level. Do we have to duplicate it ? Can we move it ?  
- [ ] Add support for ALPN in session save/load
- [ ] Enforce same ALPN when acception early data for a PSK established via a NewSessionTicket message.

Testing:
An MbedTLS client resumes a session with an MbedTLS server with a ticket associated with an ALPN not configured on the server side and indicating early data. The PSK based handshake completes successfully but the server rejects early data.

Depends on: #6341

### PR DESCRIPTION
**Pull request by waleed-elmelegy-arm**
## Description

Check if new session has the same  ALPN  as old session when receiving early data and if not reject early data.
fix #6355 
Depends on #8858

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [X] **changelog** provided
- [x] **backport** provided at #8931 
- [X] **tests** provided

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
This has the "needs-preceding-pr" label but no mention of what it depends on. Also it has a merge conflict.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -5428,7 +5439,7 @@ static int ssl_context_load(mbedtls_ssl_context *ssl,
             /* alpn_chosen should point to an item in the configured list */
             for (cur = ssl->conf->alpn_list; *cur != NULL; cur++) {
                 if (strlen(*cur) == alpn_len &&
-                    memcmp(p, cur, alpn_len) == 0) {
+                    memcmp(p, *cur, alpn_len) == 0) {
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:5442]: Good to fix that but please do it in a separate commit and add a bugfix change log. We probably need a backport for that.

File: library/ssl_tls13_server.c [hunk 2]
```diff
@@ -1839,6 +1845,21 @@ static int ssl_tls13_check_early_data_requirements(mbedtls_ssl_context *ssl)
...
+    if (alpn == NULL && ssl->session_negotiate->ticket_alpn == NULL) {
+        return 0;
+    }
+
+    if (alpn == NULL || ssl->session_negotiate->ticket_alpn == NULL
+        || memcmp(alpn, ssl->session_negotiate->ticket_alpn, strlen(alpn)) != 0) {
+        MBEDTLS_SSL_DEBUG_MSG(1, ("EarlyData: rejected, the selected ALPN is different "
+                                  "from the one associated with the pre-shared key."));
+        return -1;
+    }
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1860]:
```suggestion
    const char *alpn = mbedtls_ssl_get_alpn_protocol(ssl);
    size_t alpn_len = strlen(alpn);

    if (alpn == NULL && ssl->session_negotiate->ticket_alpn == NULL) {
        return 0;
    }

    if (alpn == NULL ||
        ssl->session_negotiate->ticket_alpn == NULL ||
        alpn_len != strlen(ssl->session_negotiate->ticket_alpn) ||
        (memcmp(alpn, ssl->session_negotiate->ticket_alpn, alpn_len) != 0)) {
        MBEDTLS_SSL_DEBUG_MSG(1, ("EarlyData: rejected, the selected ALPN is different "
                                  "from the one associated with the pre-shared key."));
        return -1;
    }
```

File: tests/include/test/ssl_helpers.h [hunk 3]
```diff
@@ -118,6 +118,9 @@ typedef struct mbedtls_test_handshake_test_options {
 #if defined(MBEDTLS_SSL_CACHE_C)
     mbedtls_ssl_cache_context *cache;
 #endif
+#if defined(MBEDTLS_SSL_ALPN)
+    const char *alpn_list[MBEDTLS_SSL_MAX_ALPN_LIST_LEN/MBEDTLS_SSL_MAX_ALPN_NAME_LEN];
```
Comment (ronald-cron-arm) [inline tests/include/test/ssl_helpers.h:122]: That's a lot, in ssl_client2/server2 it is limited to 10 (ALPN_LIST_SIZE). We should probably do something similar here.

File: library/ssl_tls.c [hunk 4]
```diff
@@ -275,6 +275,17 @@ int mbedtls_ssl_session_copy(mbedtls_ssl_session *dst,
 
 #endif /* MBEDTLS_X509_CRT_PARSE_C */
 
+#if defined(MBEDTLS_SSL_SRV_C) && defined(MBEDTLS_SSL_ALPN) && \
+    defined(MBEDTLS_SSL_EARLY_DATA)
+    {
+        int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+        ret = mbedtls_ssl_session_set_ticket_alpn(dst, src->ticket_alpn);
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:282]: Missing `dst->ticket_alpn = NULL` at the beginning of the function (see what is done for `dst->hostname`) leading to a dangling pointer it seems.

File: tests/suites/test_suite_ssl.function [hunk 5]
```diff
@@ -17,6 +17,9 @@
 #define TEST_EARLY_DATA_NO_INDICATION_SENT 1
 #define TEST_EARLY_DATA_SERVER_REJECTS 2
 #define TEST_EARLY_DATA_HRR 3
+#define TEST_EARLY_DATA_SAME_ALPN 4
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_ssl.function:20]:
```suggestion
#define TEST_EARLY_DATA_SAME_ALPN 4
```
I'd suggest to remove that scenario and just define some ALPNs if MBEDTLS_SSL_ALPN is enabled as part of TEST_EARLY_DATA_ACCEPTED.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/suites/test_suite_ssl.function (hunk 5 above)
Comment (ronald-cron-arm) [reply tests/suites/test_suite_ssl.function:20]: Updated, I suggest to remove it actually.

Comment (waleed-elmelegy-arm) [reply tests/suites/test_suite_ssl.function:20]: I added it so that we can test if no ALPN is negotiated we don't do the check.

File: ChangeLog.d/fix-alpn-negotiating-bug.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Bugfix
+   * Fix bug when comparing strings during ALPN negotiating
```
Comment (ronald-cron-arm) [inline ChangeLog.d/fix-alpn-negotiating-bug.txt:2]:
```suggestion
   * Fix the restoration of the ALPN when loading serialized connection with
   * the mbedtls_ssl_context_load() API.
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. The change log needs to be improved though (I've proposed something). Otherwise a few other suggestions for improvement.

File: library/ssl_tls13_server.c [hunk 7]
```diff
@@ -1846,6 +1845,28 @@ static int ssl_tls13_check_early_data_requirements(mbedtls_ssl_context *ssl)
...
+
+    if (alpn == NULL ||
+        ssl->session_negotiate->ticket_alpn == NULL ||
+        alpn_len != strlen(ssl->session_negotiate->ticket_alpn) ||
+        (memcmp(alpn, ssl->session_negotiate->ticket_alpn, alpn_len) != 0)) {
+        MBEDTLS_SSL_DEBUG_MSG(1, ("EarlyData: rejected, the selected ALPN is different "
+                                  "from the one associated with the pre-shared key."));
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (waleed-elmelegy-arm) [reply library/ssl_tls13_server.c:1867]: This will cause the failure to be silent and not print a failure message in case there is not ALPN in resumption.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, thanks. We need to backport the commit "Fix bug in ALPN negotiating".

File: tests/suites/test_suite_ssl.function [hunk 8]
```diff
@@ -3720,6 +3732,19 @@ void tls13_read_early_data(int scenario)
...
+#if defined(MBEDTLS_SSL_ALPN)
+    switch (scenario) {
+        case TEST_EARLY_DATA_SAME_ALPN:
+        case TEST_EARLY_DATA_DIFF_ALPN:
+        case TEST_EARLY_DATA_NO_LATER_ALPN:
+            client_options.alpn_list[0] = "ALPNExample";
+            client_options.alpn_list[1] = NULL;
+            server_options.alpn_list[0] = "ALPNExample";
+            server_options.alpn_list[1] = NULL;
+            break;
```
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_ssl.function:3744]: Should we have tests with two entries in each of the `alpn_list`s?

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM, one comment, not a blocker

File: tests/suites/test_suite_ssl.function (hunk 8 above)
Comment (ronald-cron-arm) [reply tests/suites/test_suite_ssl.function:3744]: This would rather be a test of the determination of the ALPN during the handshakes than a test that we check well the ticket ALPN against the ALPN of the resumption handshake.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM - thanks
